pkg/monitortests/client/api_unreachable.go

```diff
@@ -0,0 +1,164 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	routeclient "github.com/openshift/client-go/route/clientset/versioned"
+	utilmetrics "github.com/openshift/library-go/test/library/metrics"
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	"github.com/openshift/origin/pkg/monitortestframework"
+	"github.com/openshift/origin/pkg/monitortests/metrics"
+	"github.com/openshift/origin/pkg/test/ginkgo/junitapi"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	prometheustypes "github.com/prometheus/common/model"
+)
+
+const (
+	MonitorName = "api-unreachable-from-client"
+)
+
+func NewAPIUnReachableFromClientMonitorTest() monitortestframework.MonitorTest {
+	return &APIUnReachableFromClientMonitorTest{}
+}
+
+type monitor struct {
+	query    metrics.QueryRunner
+	analyzer metrics.SeriesAnalyzer
+	builder  *intervalBuilder
+}
+
+type APIUnReachableFromClientMonitorTest struct {
+	monitor *monitor
+}
+
+func (mt *APIUnReachableFromClientMonitorTest) StartCollection(ctx context.Context, adminRESTConfig *rest.Config, recorder monitorapi.RecorderWriter) error {
+	kubeClient, err := kubernetes.NewForConfig(adminRESTConfig)
+	if err != nil {
+		return err
+	}
+	routeClient, err := routeclient.NewForConfig(adminRESTConfig)
+	if err != nil {
+		return err
+	}
+
+	client, err := utilmetrics.NewPrometheusClient(ctx, kubeClient, routeClient)
+	if err != nil {
+		return err
+	}
+
+	kubeSvc, err := kubeClient.CoreV1().Services(metav1.NamespaceDefault).Get(ctx, "kubernetes", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster IP from kubernetes.default.svc - %v", err)
+	}
+
+	mt.monitor = &monitor{
+		query: &metrics.PrometheusQueryRunner{
+			Client:      client,
+			QueryString: `sum(rate(rest_client_requests_total{code="<error>"}[1m])) by(host)`,
+			Step:        time.Minute,
```

Comment:
I tried to explain this in the godoc:
```
The intervals are scraped from metrics, so they don't have the same granularity, as other intervals, since:
a) in OpenShift, metrics are scraped every 30s
b) for rate to be calculated, we need at lease two samples
```

If an api unreachable interval overlaps with an apiserver shutdown window, it is typically indicative of network issues at the load balancer layer. Since the intervals are grouped by host, we can also narrow it down to a particular host, for example, we have seen cases where connections over internal load balancer to be faulty at times while the service network operated just fine.
Let me know your thoughts.